src/olmo_core/distributed/parallel/__init__.py

```diff
@@ -491,23 +514,65 @@ def get_pp_stage_mesh(device_mesh: DeviceMesh, pp_mesh: Optional[DeviceMesh] = N
 
 
 def _flatten_dims(
-    device_mesh: DeviceMesh, *dims: str, name: Optional[str] = None
+    device_mesh: DeviceMesh,
+    *dims: str,
+    name: Optional[str] = None,
+    dim_names: Optional[Tuple[str, ...]] = None,
 ) -> Tuple[DeviceMesh, Tuple[str, ...]]:
+    """
+    Flatten *dims* into a single dimension called *name*.
+
+    :param device_mesh: The world-mesh object. Only views of *device_mesh* are actually mutated.
+    :param dims: The existing dimension names to merge.
+    :param name: New dimension name. If ``None`` we join *dims* with "_".
+    :param dim_names: Optional cached list of current dimension names. Supplying this avoids
+        relying on ``device_mesh.mesh_dim_names`` (which is stale after a prior
+        flatten) and therefore allows chaining multiple flatten operations.
+
+    :returns: The root mesh (now indexable by the new dimension names
+        as well as the original names) and the new dimension names.
+    """
     if name is None:
         name = "_".join(dims)
+
+    curr_names = list(dim_names or device_mesh.mesh_dim_names or [])
+    if not curr_names:
+        raise RuntimeError("Could not determine current dimension names for flattening")
+
     log.info(f"Flattening mesh dimensions {dims} into {name}")
-    assert (names := device_mesh.mesh_dim_names) is not None
-    out_names = []
-    for n in names:
+
+    out_names: list[str] = []
+    for n in curr_names:
         if n in dims:
             if name not in out_names:
                 out_names.append(name)
         else:
             out_names.append(n)
-    device_mesh[dims]._flatten(mesh_dim_name=name)
+
+    flatten_mesh(device_mesh[dims], name)  # in-place flatten on sub-mesh
     new_names = tuple(out_names)
-    return device_mesh[new_names], new_names
```

Comment:
The main functional change here is to not return a sub-mesh here indexed based on all of the new_names. Instead the root mesh is returned and the new_names are tracked separately. This has the annoying property that `device_mesh.mesh_dim_names` does not reflect new_names. But new_names can still be used to index into the root device_mesh.